<h2>Data Analytics Homework 2 - David Mallon (08597596) & Paul O'Donovan (22208104)</h2>

Prepare a data quality report for your CSV file. 

In [68]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [69]:
# Reading from a csv file, into a data frame
df1 = pd.read_csv('covid19-cdc-22208104.csv', keep_default_na=True, delimiter=',', engine='python', skipinitialspace=True)

# Reading from a csv file, into a data frame
df2 = pd.read_csv('covid19-cdc-8597596.csv', keep_default_na=True, delimiter=',', engine='python', skipinitialspace=True)

# Concatenate the data frames
df = pd.concat([df1, df2], axis=0)

In [70]:
# Check how many rows and columns this dataframe has
print("Number of rows and columns:", df.shape)

# Show first 5 rows of data frame
# The rows are indexed starting from 0
df.head(5)

Number of rows and columns: (40000, 19)


case_month res_state  state_fips_code    res_county  county_fips_code  \
0    2021-04        IL             17.0          COOK           17031.0   
1    2020-04        MA             25.0       SUFFOLK           25025.0   
2    2020-11        PA             42.0  PHILADELPHIA           42101.0   
3    2021-12        NY             36.0        QUEENS           36081.0   
4    2022-01        AZ              4.0      MARICOPA            4013.0   

        age_group     sex   race            ethnicity  \
0  50 to 64 years    Male  White      Hispanic/Latino   
1       65+ years    Male  White  Non-Hispanic/Latino   
2       65+ years  Female  White              Unknown   
3       65+ years  Female  Asian  Non-Hispanic/Latino   
4       65+ years    Male  White  Non-Hispanic/Latino   

   case_positive_specimen_interval  case_onset_interval  process exposure_yn  \
0                              1.0                  0.0  Missing     Missing   
1                              0.0                  NaN  Missing     Missing   
2                              0.0                  NaN  Missing         Yes   
3                              0.0                  NaN  Missing     Missing   
4                              NaN                  NaN  Missing     Missing   

              current_status symptom_status  hosp_yn   icu_yn death_yn  \
0  Laboratory-confirmed case        Missing      Yes  Missing      Yes   
1  Laboratory-confirmed case        Missing      Yes  Missing      Yes   
2  Laboratory-confirmed case    Symptomatic  Unknown  Unknown      Yes   
3  Laboratory-confirmed case        Unknown      Yes      Yes      Yes   
4              Probable Case        Missing  Missing  Missing      Yes   

  underlying_conditions_yn  
0                      NaN  
1                      NaN  
2                      NaN  
3                      Yes  
4                      NaN

In [71]:
#Get a summary of columns
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40000 entries, 0 to 19999
Data columns (total 19 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   case_month                       40000 non-null  object 
 1   res_state                        39999 non-null  object 
 2   state_fips_code                  39999 non-null  float64
 3   res_county                       37702 non-null  object 
 4   county_fips_code                 37702 non-null  float64
 5   age_group                        39740 non-null  object 
 6   sex                              39234 non-null  object 
 7   race                             35303 non-null  object 
 8   ethnicity                        34875 non-null  object 
 9   case_positive_specimen_interval  20865 non-null  float64
 10  case_onset_interval              17552 non-null  float64
 11  process                          40000 non-null  object 
 12  exposure_yn       

In [72]:
df.describe().T

count          mean           std     min  \
state_fips_code                  39999.0     29.556239     13.341753     1.0   
county_fips_code                 37702.0  29578.939605  13211.869250  1001.0   
case_positive_specimen_interval  20865.0      0.194201      2.312371   -90.0   
case_onset_interval              17552.0     -0.057714      1.800305   -60.0   

                                     25%      50%      75%      max  
state_fips_code                     19.0     34.0     37.0     78.0  
county_fips_code                 18163.0  34017.0  37183.0  56039.0  
case_positive_specimen_interval      0.0      0.0      0.0     94.0  
case_onset_interval                  0.0      0.0      0.0     69.0

Per the above results, I am missing 2298 pieces of information for county_fips_code. This is fine in my opinion as I have the almost full 39999 pieces of information for state_fips_code which will suffice for the purposes of this examination. The county_fips_code is not crucial and is sufficiently replaced and supplemented by the state_fips_code data.

In relation to the missing / blank data for case_positive_specimen_interval, I will consider both sets of data in section 2 (my data quality plan) as I will consider their inclusion, possible amendments/corrections and any links that can be made with other provided information.

In [73]:
# Look at column names. Ensure there are not spaces in or after the name
df.columns

Index(['case_month', 'res_state', 'state_fips_code', 'res_county',
       'county_fips_code', 'age_group', 'sex', 'race', 'ethnicity',
       'case_positive_specimen_interval', 'case_onset_interval', 'process',
       'exposure_yn', 'current_status', 'symptom_status', 'hosp_yn', 'icu_yn',
       'death_yn', 'underlying_conditions_yn'],
      dtype='object')

Neither the state_fips_code and county_fips_code data represents a continuous range of values, but instead represent specific states and counties respectively, which are both categorical attributes. Both these datasets are better treated as categorical variables, which will help us understand the distribution of the data and any relationships or patterns that may exist between the state codes and other variables in the dataset.

In [74]:
df["state_fips_code"] = df["state_fips_code"].astype("object")
df["county_fips_code"] = df["county_fips_code"].astype("object")

#Select all columns of type 'object'
object_columns = df.select_dtypes(['object']).columns

#Convert selected columns to type 'category' purely for ease of reference when reviewing categorical and continuous characteristics.
for column in object_columns:
    df[column] = df[column].astype('category')

df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40000 entries, 0 to 19999
Data columns (total 19 columns):
 #   Column                           Non-Null Count  Dtype   
---  ------                           --------------  -----   
 0   case_month                       40000 non-null  category
 1   res_state                        39999 non-null  category
 2   state_fips_code                  39999 non-null  category
 3   res_county                       37702 non-null  category
 4   county_fips_code                 37702 non-null  category
 5   age_group                        39740 non-null  category
 6   sex                              39234 non-null  category
 7   race                             35303 non-null  category
 8   ethnicity                        34875 non-null  category
 9   case_positive_specimen_interval  20865 non-null  float64 
 10  case_onset_interval              17552 non-null  float64 
 11  process                          40000 non-null  category
 12  expo

Convert the features to their appropriate data types (e.g., decide which features are more appropriate as continuous and which ones as categorical types).

Look for duplicate rows and columns. Consider whether it makes sense to keep them or drop them.

In [75]:
#Print the number of duplicates, without the original rows that were duplicated
print('Number of duplicate (excluding first original occurence) rows in the table is: ', df.duplicated().sum())

# Check for duplicate rows. 
# Use "keep=False" to mark all duplicates as true, including the original rows that were duplicated.
print('Number of duplicate rows (including first original occurence) in the table is:', df[df.duplicated(keep=False)].shape[0])

Number of duplicate (excluding first original occurence) rows in the table is:  3639
Number of duplicate rows (including first original occurence) in the table is: 6178


In [76]:
view_duplicates = df[df.duplicated(keep = False)]
view_duplicates.to_csv('duplicate_columns.csv', index=False)

Having reviewed the duplicates (as set out in a separate csv I prepared), I was satisfied that these should stay in the dataframe. As the information contained therein has been intentionally anonymised, there are not enough specific identifiers relating to each patient's data which would justify removing the duplicates. I believe it better to err on the side of caution and not remove what could (and probably is) information recorded which is identical but actually relates to different individuals, rather than remove valid information on the assumption that it is a duplicate entry.

In [77]:
# quick view of duplicates

duplicate = df[df.duplicated()]
 
print(duplicate)
 

      case_month res_state state_fips_code   res_county county_fips_code  \
235      2020-03        NY            36.0       QUEENS          36081.0   
274      2021-01        CA             6.0  LOS ANGELES           6037.0   
289      2020-12        CA             6.0  LOS ANGELES           6037.0   
373      2021-01        FL            12.0   MIAMI-DADE          12086.0   
476      2021-12        IL            17.0         COOK          17031.0   
...          ...       ...             ...          ...              ...   
19933    2020-07        TX            48.0      TARRANT          48439.0   
19934    2022-01        IN            18.0       MARION          18097.0   
19941    2021-12        NJ            34.0        OCEAN          34029.0   
19944    2022-01        IN            18.0       GIBSON          18051.0   
19972    2022-01        MN            27.0     HENNEPIN          27053.0   

            age_group      sex     race            ethnicity  \
235         65+ years  

Checking if any duplicate columns present.

Function sourced from https://www.geeksforgeeks.org/how-to-find-drop-duplicate-columns-in-a-pandas-dataframe/ 

In [78]:
def getDuplicateColumns(df):
    duplicateColumnNames = set()
 
    for x in range(df.shape[1]):
        col = df.iloc[:, x]
        for y in range(x + 1, df.shape[1]):
            otherCol = df.iloc[:, y]
            if col.equals(otherCol):
                duplicateColumnNames.add(df.columns.values[y])
    return list(duplicateColumnNames)

duplicateColNames = getDuplicateColumns(df)

for column in duplicateColNames:
    print('Column Name : ', column)

In [79]:
# Check for duplicate columns - additional test
#First transpose the df so columns become rows, then apply the same check as above
dfT = df.T
print("Number of duplicate (excluding first) columns in the table is: ", dfT.duplicated().sum())
print("Number of duplicate (including first) columns in the table is: ",  dfT[dfT.duplicated(keep=False)].shape[0])

Number of duplicate (excluding first) columns in the table is:  0
Number of duplicate (including first) columns in the table is:  0


Look for constant columns. Consider whether it makes sense to keep them or drop them.

In [80]:
# Check the number of unique values in each column
nunique_values = df.nunique()

constant_columns = nunique_values[nunique_values == 1].index
constant_column_names = list(constant_columns)

# Print the names of the constant columns
print("Constant columns: ", constant_column_names)

Constant columns:  []


In [81]:
c = [c for c in df.columns if len(set(df[c])) == 1]
print ("Constant columns: ", c)

Constant columns:  []


There are no 100% constant columns present in my dataset, therefore no adjustment is required.
I have rechecked the data basing it on a 90% occurrence (when excluding empty cells) for completeness per the below cells and I would opine that the information in both the 'process' and 'underlying_conditions_yn' cells warrant inclusion in the cleaned dataframe due to weighing up the results against the number of empty cells with missing information. Further examination of these features is undertaken in part two of this assignment.

In [82]:
#Calculate the percentage of each value in each column, including empty cells
value_percentages = df.apply(lambda x: x.value_counts(normalize=True, dropna=False).max())

#Select only the columns where the maximum value percentage is >= 0.95
constant_columns = value_percentages[value_percentages >= 0.90].index

#List the names of the constant columns
constant_column_names = list(constant_columns)

#Print the names of the constant columns
print("Constant columns: ", constant_column_names)

Constant columns:  ['process', 'underlying_conditions_yn']


The information from both 'process' and 'underlying_conditions_yn' is still usefull for consideration based on the above commentary. This information will also be useful when preparing my data quality plan and expected results from the bar plots. Further commentary is provided on my logic is provided in both my Data Quality Report and Data Quality Plan.

In [83]:
df['process'].value_counts()

Missing                            36463
Clinical evaluation                 1663
Laboratory reported                  730
Routine surveillance                 460
Multiple                             292
Contact tracing of case patient      134
Unknown                              113
Provider reported                     80
Other                                 64
Autopsy                                1
Name: process, dtype: int64

In [84]:
df['underlying_conditions_yn'].value_counts()

Yes    3444
No       51
Name: underlying_conditions_yn, dtype: int64

# Q. Save your updated/cleaned data frame to a new csv file.

In [85]:
df.to_csv('cleaned_data_hw2.csv', index=False)

In [87]:
# Reading cleaned data into a data frame
df = pd.read_csv('cleaned_data_hw2.csv')
df["state_fips_code"] = df["state_fips_code"].astype("object")
df["county_fips_code"] = df["county_fips_code"].astype("object")

#Select all columns of type 'object'
object_columns = df.select_dtypes(['object']).columns

#Convert selected columns to type 'category' purely for ease of reference when reviewing categorical and continuous characteristics.
for column in object_columns:
    df[column] = df[column].astype('category')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 19 columns):
 #   Column                           Non-Null Count  Dtype   
---  ------                           --------------  -----   
 0   case_month                       40000 non-null  category
 1   res_state                        39999 non-null  category
 2   state_fips_code                  39999 non-null  category
 3   res_county                       37702 non-null  category
 4   county_fips_code                 37702 non-null  category
 5   age_group                        39740 non-null  category
 6   sex                              39234 non-null  category
 7   race                             35303 non-null  category
 8   ethnicity                        34875 non-null  category
 9   case_positive_specimen_interval  20865 non-null  float64 
 10  case_onset_interval              17552 non-null  float64 
 11  process                          40000 non-null  category
 12  expo